x/fees/keeper/msg_server.go

```diff
@@ -0,0 +1,144 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/tharsis/evmos/v3/x/fees/types"
+)
+
+var _ types.MsgServer = &Keeper{}
+
+// RegisterDevFeeInfo registers a contract to receive transaction fees
+func (k Keeper) RegisterDevFeeInfo(
+	goCtx context.Context,
+	msg *types.MsgRegisterDevFeeInfo,
+) (*types.MsgRegisterDevFeeInfoResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if !k.isEnabled(ctx) {
+		return nil, sdkerrors.Wrapf(types.ErrInternalFee, "fees module is not enabled")
+	}
+
+	contract := common.HexToAddress(msg.ContractAddress)
+	if k.IsFeeRegistered(ctx, contract) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "contract %s is already registered", contract)
+	}
+
+	var withdrawal *sdk.AccAddress
+	deployer, _ := sdk.AccAddressFromBech32(msg.DeployerAddress)
+	derivedContractAddr := common.BytesToAddress(deployer)
+
+	if msg.WithdrawAddress != "" {
+		_withdrawal, _ := sdk.AccAddressFromBech32(msg.WithdrawAddress)
+		withdrawal = &_withdrawal
+	}
+
+	for _, nonce := range msg.Nonces {
+		derivedContractAddr = crypto.CreateAddress(derivedContractAddr, nonce)
+	}
+
+	if contract != derivedContractAddr {
+		return nil, sdkerrors.Wrapf(
+			sdkerrors.ErrorInvalidSigner,
+			"%s not contract deployer or wrong nonce", msg.DeployerAddress,
```

Comment:
error message should contain the expected contract and the derived contract addresses